File: frontend/src/types/employee.ts
export type Employee = {
  id: number
  first_name: string
  last_name: string
  full_name: string
  job_title: string
  country: string
  salary: string | null
  email: string
  created_at?: string
  updated_at?: string
}

export type PaginationMeta = {
  page: number
  pages: number
  count: number
  items: number
  from: number
  to: number
}

export type PaginatedEmployees = {
  data: Employee[]
  meta: PaginationMeta
}

export type EmployeeInput = {
  first_name: string
  last_name: string
  job_title: string
  country: string
  email: string
  salary: string
}

File: app/models/employee.rb
# == Schema Information
#
# Table name: employees
#
#  id         :integer          not null, primary key
#  first_name :string
#  last_name  :string
#  job_title  :string
#  salary     :decimal(12, 2)
#  country    :string
#  email      :string
#  created_at :datetime         not null
#  updated_at :datetime         not null
#
class Employee < ApplicationRecord
  scope :search, lambda { |query|
    q = query.to_s.strip
    if q.blank?
      all
    else
      pattern = "%#{Employee.sanitize_sql_like(q.downcase)}%"
      where(
        <<~SQL.squish,
          LOWER(first_name) LIKE :p OR LOWER(last_name) LIKE :p OR
          LOWER(email) LIKE :p OR LOWER(job_title) LIKE :p OR LOWER(country) LIKE :p
        SQL
        p: pattern
      )
    end
  }
end
